src/hotspot/share/gc/g1/g1StringDedup.hpp

```diff
@@ -75,7 +75,8 @@ class G1StringDedup : public StringDedup {
   // Enqueues a deduplication candidate for later processing by the deduplication
   // thread. Before enqueuing, these functions apply the appropriate candidate
   // selection policy to filters out non-candidates.
-  static void enqueue_from_mark(oop java_string, uint worker_id);
+  static void enqueue_from_mark(oop obj, uint worker_id);
```

Comment:
As Albert mentioned, it would be nice if the caller of this would also get this optimization as it seems straightforward to implement there as well for consistency.

Reply:
Because of compressed klasses, getting the klass is never completely trivial.  However, since effectively the first thing done by enqueue_from_mark is to test for the obj being a String, hoisting that out to fullgc's mark_object seems reasonable.  And that gives the consistency between the enqueue functions that Albert asked for.  I'll add that to the PR after re-running tests.